cpp/doc/DOCUMENTATION.md

```diff
@@ -0,0 +1,574 @@
+ */
+```
+
+Doxygen comment blocks start with `/**` and end with `*/` only, and with nothing else on those
+lines. Do not add dashes `-----` or extra asterisks `*****` to the first and last lines of a doxygen
+block. The block must be placed immediately before the source code line to which it refers. The
+block may be indented to line up vertically with the item it documents as appropriate. See the
+[Example](#the-example) section below.
+
+Each line in the comment block between the `/**` and `*/` lines should start with a space followed
+by an asterisk. Any text on these lines, including tag declarations, should start after a single
+space after the asterisk.
+
+## Tag/Command names
+
+Use `@` to prefix doxygen commands (e.g. `@brief`, `@code`, etc.)
+
+## Markdown
+
+The doxygen tool supports a limited set of markdown format in the comment block including links,
+tables, lists, etc. In some cases a trade-off may be required for readability in the source text
+file versus the readability in the doxygen formatted web pages. For example, there are some
+limitations on readability with '%' character and pipe character '|' within a markdown table.
+
+Avoid using direct HTML tags. Although doxygen supports markdown and markdown supports HTML tags,
+the HTML support for doxygen's markdown is also limited.
+
+## The Example
+
+The following example covers most of the doxygen block comment and tag styles for documenting C++
+code in libcuspatial.
+
+```c++
+/**
+ * @file source_file.cpp
+ * @brief Description of source file contents
+ *
+ * Longer description of the source file contents.
+ */
+
+/**
+ * @brief One sentence description of the class.
+ *
+ * @ingroup optional_predefined_group_id
+ *
+ * Longer, more detailed description of the class.
+ *
+ * @tparam T Short description of each template parameter
+ * @tparam U Short description of each template parameter
+ */
+template <typename T, typename U>
+class example_class {
+
+  void get_my_int();            ///< Simple members can be documented like this
+  void set_my_int( int value ); ///< Try to use descriptive member names
+
+  /**
+   * @brief Short, one sentence description of the member function.
+   *
+   * A more detailed description of what this function does and what
+   * its logic does.
+   *
+   * @code
+   * example_class<int> inst;
+   * inst.set_my_int(5);
+   * int output = inst.complicated_function(1,dptr,fptr);
+   * @endcode
+   *
+   * @param[in]     first  This parameter is an input parameter to the function
+   * @param[in,out] second This parameter is used both as an input and output
+   * @param[out]    third  This parameter is an output of the function
+   *
+   * @return The result of the complex function
+   */
+  T complicated_function(int first, double* second, float* third)
+  {
+      // Do not use doxygen-style block comments
+      // for code logic documentation.
+  }
+
+ private:
+  int my_int;                ///< An example private member variable
```

Comment:
Would be nice to provide an example for "between the line" doxygen comments, like
```suggestion
  int my_int;                ///< An example private member variable
  /// Another example of private member variable
  float my_float;
```
Not that single line comment will automatically be treated as @brief and mutliple line comments are viewed as detail.